### ISSUE TITLE
Inconsistent behavior when running on Windows or Ubuntu 22.04 (inside Docker) and using ed25519 private key

### ISSUE THREAD

**Issue by flelli** [mwiede/jsch#316]
Running the same code, with the same ed25519  key passed as a string drives to a different behavior when running on Windows or Linux.

Debug on Windows is:

```
Connecting to github.com port 22
Connection established
Remote version string: SSH-2.0-babeld-8f627ebe
Local version string: SSH-2.0-JSCH_0.2.7
CheckCiphers: chacha20-poly1305@openssh.com
CheckKexes: curve25519-sha256,curve25519-sha256@libssh.org,curve448-sha512
CheckSignatures: ssh-ed25519,ssh-ed448
server_host_key proposal before known_host reordering is: ssh-ed25519,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-512,rsa-sha2-256
server_host_key proposal after known_host reordering is: ssh-ed25519,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-512,rsa-sha2-256
SSH_MSG_KEXINIT sent
SSH_MSG_KEXINIT received
kex: server: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256
kex: server: ssh-ed25519,ecdsa-sha2-nistp256,rsa-sha2-512,rsa-sha2-256,ssh-rsa
kex: server: chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr
kex: server: chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr
kex: server: hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512,hmac-sha2-256
kex: server: hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512,hmac-sha2-256
kex: server: none,zlib@openssh.com,zlib
kex: server: none,zlib@openssh.com,zlib
kex: server: 
kex: server: 
kex: client: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha256,ext-info-c
kex: client: ssh-ed25519,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-512,rsa-sha2-256
kex: client: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
kex: client: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
kex: client: hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
kex: client: hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
kex: client: none,zlib@openssh.com,zlib
kex: client: none,zlib@openssh.com,zlib
kex: client: 
kex: client: 
kex: algorithm: curve25519-sha256
kex: host key algorithm: ssh-ed25519
kex: server->client cipher: aes128-ctr MAC: hmac-sha2-256-etm@openssh.com compression: none
kex: client->server cipher: aes128-ctr MAC: hmac-sha2-256-etm@openssh.com compression: none
SSH_MSG_KEX_ECDH_INIT sent
expecting SSH_MSG_KEX_ECDH_REPLY
ssh_eddsa_verify: ssh-ed25519 signature true
Permanently added 'github.com' (EDDSA) to the list of known hosts.
SSH_MSG_NEWKEYS sent
SSH_MSG_NEWKEYS received
SSH_MSG_SERVICE_REQUEST sent
SSH_MSG_EXT_INFO received
server-sig-algs=<ssh-ed25519-cert-v01@openssh.com,ecdsa-sha2-nistp521-cert-v01@openssh.com,ecdsa-sha2-nistp384-cert-v01@openssh.com,ecdsa-sha2-nistp256-cert-v01@openssh.com,sk-ssh-ed25519-cert-v01@openssh.com,sk-ecdsa-sha2-nistp256-cert-v01@openssh.com,rsa-sha2-512-cert-v01@openssh.com,rsa-sha2-256-cert-v01@openssh.com,ssh-rsa-cert-v01@openssh.com,sk-ssh-ed25519@openssh.com,sk-ecdsa-sha2-nistp256@openssh.com,ssh-ed25519,ecdsa-sha2-nistp521,ecdsa-sha2-nistp384,ecdsa-sha2-nistp256,rsa-sha2-512,rsa-sha2-256,ssh-rsa>
SSH_MSG_SERVICE_ACCEPT received
Authentications that can continue: publickey
Next authentication method: publickey
PubkeyAcceptedAlgorithms = ssh-ed25519,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-512,rsa-sha2-256
PubkeyAcceptedAlgorithms in server-sig-algs = [ssh-ed25519, ecdsa-sha2-nistp256, ecdsa-sha2-nistp384, ecdsa-sha2-nistp521, rsa-sha2-512, rsa-sha2-256]
ssh-ed25519 preauth success
ssh-ed25519 auth success
Authentication succeeded (publickey).
Disconnecting from github.com port 22
Caught an exception, leaving main loop due to Socket closed
```

While on Ubuntu 22.04 is:

```
Connecting to github.com port 22
Connection established
Remote version string: SSH-2.0-babeld-8f627ebe
Local version string: SSH-2.0-JSCH_0.2.7
CheckCiphers: chacha20-poly1305@openssh.com
CheckKexes: curve25519-sha256,curve25519-sha256@libssh.org,curve448-sha512
CheckSignatures: ssh-ed25519,ssh-ed448
server_host_key proposal before known_host reordering is: ssh-ed25519,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-512,rsa-sha2-256
server_host_key proposal after known_host reordering is: ssh-ed25519,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-512,rsa-sha2-256
SSH_MSG_KEXINIT sent
SSH_MSG_KEXINIT received
kex: server: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256
kex: server: ssh-ed25519,ecdsa-sha2-nistp256,rsa-sha2-512,rsa-sha2-256,ssh-rsa
kex: server: chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr
kex: server: chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr
kex: server: hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512,hmac-sha2-256
kex: server: hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512,hmac-sha2-256
kex: server: none,zlib@openssh.com,zlib
kex: server: none,zlib@openssh.com,zlib
kex: server: 
kex: server: 
kex: client: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha256,ext-info-c
kex: client: ssh-ed25519,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-512,rsa-sha2-256
kex: client: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
kex: client: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
kex: client: hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
kex: client: hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
kex: client: none,zlib@openssh.com,zlib
kex: client: none,zlib@openssh.com,zlib
kex: client: 
kex: client: 
kex: algorithm: curve25519-sha256
kex: host key algorithm: ssh-ed25519
kex: server->client cipher: aes128-ctr MAC: hmac-sha2-256-etm@openssh.com compression: none
kex: client->server cipher: aes128-ctr MAC: hmac-sha2-256-etm@openssh.com compression: none
SSH_MSG_KEX_ECDH_INIT sent
expecting SSH_MSG_KEX_ECDH_REPLY
ssh_eddsa_verify: ssh-ed25519 signature true
Permanently added 'github.com' (EDDSA) to the list of known hosts.
SSH_MSG_NEWKEYS sent
SSH_MSG_NEWKEYS received
SSH_MSG_SERVICE_REQUEST sent
SSH_MSG_EXT_INFO received
server-sig-algs=<ssh-ed25519-cert-v01@openssh.com,ecdsa-sha2-nistp521-cert-v01@openssh.com,ecdsa-sha2-nistp384-cert-v01@openssh.com,ecdsa-sha2-nistp256-cert-v01@openssh.com,sk-ssh-ed25519-cert-v01@openssh.com,sk-ecdsa-sha2-nistp256-cert-v01@openssh.com,rsa-sha2-512-cert-v01@openssh.com,rsa-sha2-256-cert-v01@openssh.com,ssh-rsa-cert-v01@openssh.com,sk-ssh-ed25519@openssh.com,sk-ecdsa-sha2-nistp256@openssh.com,ssh-ed25519,ecdsa-sha2-nistp521,ecdsa-sha2-nistp384,ecdsa-sha2-nistp256,rsa-sha2-512,rsa-sha2-256,ssh-rsa>
SSH_MSG_SERVICE_ACCEPT received
Authentications that can continue: publickey
Next authentication method: publickey
Disconnecting from github.com port 22
```

As you can see the logs are poerfectly the same but for the last lines, after `Next authentication method: publickey`, which shows succesful lines on Windows and nothing on Ubuntu.

Another weird fact is that running on GitHub Actions, using Ubuntu 22.04 as well, things work.

Any help?

Thanks

**Comment by flelli** [comment]
JSch version is 0.2.8

**Comment by norrisjeremy** [comment]
Hi @flelli,

Unfortunately I am unable to replicate the issue you have reported, both on an Ubuntu 22.04 workstation, as well as with an Ubuntu 22.04 Docker container:
```
import com.jcraft.jsch.JSch;
import com.jcraft.jsch.Logger;

public class JSch316 {
  public static void main(String[] args) throws Exception {
    if (args.length != 2) {
      throw new IllegalArgumentException("missing args");
    }

    var key = args[0];
    var passphrase = args[1];

    JSch.setLogger(new SystemOutLogger());
    var jsch = new JSch();
    jsch.addIdentity(key, passphrase);
    var session = jsch.getSession("git", "github.com");
    session.setConfig("StrictHostKeyChecking", "no");
    session.connect();
    var channel = session.openChannel("shell");
    session.disconnect();
  }

  public static class SystemOutLogger implements Logger {
    @Override
    public boolean isEnabled(int level) {
      return true;
    }

    @Override
    public void log(int level, String message) {
      System.out.println(message);
    }

    @Override
    public void log(int level, String message, Throwable cause) {
      System.out.println(message);
      cause.printStackTrace();
    }
  }
}
```

I've executed the above code using: `java -cp jsch-0.2.8.jar JSch316.java ~/.ssh/id_ed25519 <mypassphrase>` and it works in both cases.

Can you provide more details or a way consistently reproduce?

Thanks,
Jeremy

**Comment by flelli** [comment]
Hi @norrisjeremy ,

thanks a lot for your reply. I tried running your same snipped in the dev container that is giving me issues and... it works. I also adapted it to my specific use case (i.e. passing the private key as a string, with or without passphrase) and keeps working.

It would be really helpful if I had some means to get additional logs after the `Next authentication method: publickey`, to figure out why in one case it actually does additional things and it doesn't in the other case.

Or maybe you can take a glance at [this code](https://github.com/mooltiverse/nyx/blob/main/modules/java/main/src/main/java/com/mooltiverse/oss/nyx/git/JGitRepository.java#L142), which is the one I use and gives me problems, as you may likely notice something I did wrong.

**Comment by norrisjeremy** [comment]
Hi @flelli,

I'd suggest debugging `JschConfigSessionFactory.getSession()` from JGit around where it calls JSch's `Session.connect()`.
My first guess is that JSch's `Session.connect()` is throwing an Exception that is getting swallowed or not handled by JGit.

Thanks,
Jeremy

**Comment by flelli** [comment]
Hi @norrisjeremy ,

I narrowed the issue and it's about (the way I use) the `SSHAgentConnector`. I followed [this example](https://github.com/mwiede/jsch/blob/master/examples/JSchWithAgentProxy.java) but maybe I missed something.

When the `SSHAgentConnector`  constructor throws an exception (`AgentProxyException: SSH_AUTH_SOCK is not defined`) I just don't plug it in and everything works.

But when no exception is thrown I set it up using an `AgentIdentityRepository`, then nothing happens after the `Next authentication method: publickey` log message.

In the specific case I'm testing I'm passing the key as a string with no password so the `SSHAgentConnector` should do nothing. But somehow it prevents the process to continue as if the `SSHAgentConnector` was not there.

[This](https://github.com/mooltiverse/nyx/blob/main/modules/java/main/src/main/java/com/mooltiverse/oss/nyx/git/JGitRepository.java#L245) try/catch contains the code.

Any chance you can point me to what I'm doing wrong?

Basically I need:

- the `SSHAgentConnector` to be opaque and not prevent the whole process to complete when it's been plugged in, even if not used, or
- know when I need to plug it in or not in advance

Many thanks!

**Comment by norrisjeremy** [comment]
Hi @flelli,

Do you have a functional ssh-agent running when these failures occur (that has already been loaded with a valid ssh key to utilize for authentication by executing `ssh-add`)?

Thanks,
Jeremy

**Comment by flelli** [comment]
No, not in this test scenario, although the final user may have it at runtime.

Thanks a lot @norrisjeremy !

**Comment by norrisjeremy** [comment]
Btw, I suspect your issue is that you are falsely assuming that JSch should fallback to utilizing the private key you have loaded by your earlier call to `defaultJSch.addIdentity("git", privateKey.getBytes(), null, passphrase)` after calling `defaultJSch.setIdentityRepository(irepo)` if an ssh-agent isn't available or lacks a usable ssh key for authentication to your server.

That's not the way that JSch functions however: the call `defaultJSch.setIdentityRepository(irepo)` is instructing JSch to strictly utilize an `ssh-agent` for publickey authentication; at this point, the prior call to `defaultJSch.addIdentity("git", privateKey.getBytes(), null, passphrase)` is void and has no effect (JSch does not maintain it as a "fallback" if the ssh-agent is unusable).

You should only be calling `defaultJSch.setIdentityRepository(irepo)` if you are prepared to strictly use the ssh-agent as the source of truth for publickey authentication.

**Comment by flelli** [comment]
You're totally right, I indeed though there was a fallback logic here. This explains everything.

So thanks @norrisjeremy for your hint. I'll change the code according tho this.
